clang/lib/Sema/SemaCast.cpp

```diff
@@ -2425,8 +2461,16 @@ static TryCastResult TryReinterpretCast(Sema &Self, ExprResult &SrcExpr,
         return TC_Failed;
       }
     }
-    Kind = SrcIsCap && !DestType->isIntCapType() ? CK_CHERICapabilityToAddress
-                                                 : CK_PointerToIntegral;
+    if (SrcIsCap && !DestType->isIntCapType()) {
+      Kind = CK_CHERICapabilityToAddress;
+      // XXXAR: this would be more consistent but would cause too much code to
+      //  not behave as expected in offset mode:
```

Comment:
Why is this a problem that reinterpret_cast introduces over plain C casts?

Reply:
C casts have the same issue: in offset mode `(size_t)cap_ptr` always returns an address and not an offset.
Changing that would change offset mode from "mostly compatible" to "completely incompatible" with existing C code.